Don't initialize `SERVER_PORT` in ENV to empty String

adam12 opened the pull request:
The Rack specification states that if `SERVER_PORT` is present,
it must be a numeric value.

When initializing the Hash, the `SERVER_PORT` key is set, but
never filled in.

Closes #107 

I'm not sure if this is the most optimal solution but I wanted to offer something.

boazsegev commented on ext/iodine/iodine_http.c:
```diff
@@ -365,7 +365,11 @@ static inline VALUE copy2env(iodine_http_request_handle_s *handle) {
     rb_hash_aset(env, SERVER_NAME,
                  rb_enc_str_new(tmp.data, tmp.len, IodineBinaryEncoding));
-    rb_hash_aset(env, SERVER_PORT, QUERY_ESTRING);
+    if (RSTRING_LEN(QUERY_ESTRING) != 0) {
```
This is my fault for having poorly named variables, but... the `ESTRING` in `QUERY_ESTRING` stands for an **Empty** String, its length is always zero and it's used almost everywhere I needed an empty frozen string (to avoid creating multiple objects)... so there's really no need for this `if` statement.

boazsegev commented on ext/iodine/iodine_http.c:
```diff
@@ -365,7 +365,11 @@ static inline VALUE copy2env(iodine_http_request_handle_s *handle) {
+    if (RSTRING_LEN(QUERY_ESTRING) != 0) {
+      rb_hash_aset(env, SERVER_PORT, QUERY_ESTRING);
+    } else {
+      rb_hash_delete(env, SERVER_PORT);
```
I think that using `rb_hash_aset(env, SERVER_PORT, Qnil)` would be both faster here and in the client code.

Using the `delete` operator can cause possible hash map misses when searching for a value and these misses are slower than finding a `nil` value.

This should work with the way the Linter middleware checks the value: https://github.com/rack/rack/blob/d15dd728440710cfc35ed155d66a98dc2c07ae42/lib/rack/lint.rb#L284

adam12 replied:
Unfortunately I tried `Qnil` originally, and it failed. I just tried it again to confirm.

`ERROR: Iodine caught an unprotected exception - Rack::Lint::LintError: env variable SERVER_PORT has non-string value nil`

```diff
diff --git a/ext/iodine/iodine_http.c b/ext/iodine/iodine_http.c
index 4f7475f..a9ca702 100644
--- a/ext/iodine/iodine_http.c
+++ b/ext/iodine/iodine_http.c
@@ -365,11 +365,7 @@ static inline VALUE copy2env(iodine_http_request_handle_s *handle) {
   if (*pos == 0) {
     rb_hash_aset(env, SERVER_NAME,
                  rb_enc_str_new(tmp.data, tmp.len, IodineBinaryEncoding));
-    if (RSTRING_LEN(QUERY_ESTRING) != 0) {
-      rb_hash_aset(env, SERVER_PORT, QUERY_ESTRING);
-    } else {
-      rb_hash_delete(env, SERVER_PORT);
-    }
+    rb_hash_aset(env, SERVER_PORT, Qnil);
   } else {
     rb_hash_aset(
         env, SERVER_NAME,
```

This is because the described spec is wrong, and non-String values are failures too, if they don't contain a period: https://github.com/rack/rack/blob/d15dd728440710cfc35ed155d66a98dc2c07ae42/lib/rack/lint.rb#L308-L315

adam12 replied:
Actually, I think the spec is right, they are just using the presence of `nil` to guard on the coercion through `Integer`. So it looks like it allows `nil`, but in essence they only care in that scenario that they don't coerce a `nil` value and trigger an error at that line.

My guess anyways.

adam12 replied:
I could drop the conditional and always delete. Or could we remove the empty string that is initialized earlier in this file (`rb_hash_aset(env_template_no_upgrade, SERVER_PORT, QUERY_ESTRING);`)?

```
diff --git a/ext/iodine/iodine_http.c b/ext/iodine/iodine_http.c
index 4f7475f..7f47c17 100644
--- a/ext/iodine/iodine_http.c
+++ b/ext/iodine/iodine_http.c
@@ -365,11 +365,7 @@ static inline VALUE copy2env(iodine_http_request_handle_s *handle) {
   if (*pos == 0) {
     rb_hash_aset(env, SERVER_NAME,
                  rb_enc_str_new(tmp.data, tmp.len, IodineBinaryEncoding));
-    if (RSTRING_LEN(QUERY_ESTRING) != 0) {
-      rb_hash_aset(env, SERVER_PORT, QUERY_ESTRING);
-    } else {
-      rb_hash_delete(env, SERVER_PORT);
-    }
+    rb_hash_delete(env, SERVER_PORT);
   } else {
     rb_hash_aset(
         env, SERVER_NAME,
```

boazsegev reviewed (changes requested):
Thank you for discovering this issue and providing a possible fix.

I love the general approach and its simplicity, but I think the change would have been more effective by replacing `QUERY_ESTRING` with `Qnil` in the `rb_hash_aset` function call you replaced:

```c
rb_hash_aset(env, SERVER_PORT, Qnil);
```

You can read my notes as to why.

If you update the PR that would be great, if not, I'll do it myself... eventually. I have a family crisis so I'm not really working much these days.

Thanks!

adam12 commented:
> I have a family crisis so I'm not really working much these days.

I'm sorry to hear that. I hope things improve for you.

adam12 commented:
I updated the patch to skip initializing the Hash with the `SERVER_PORT` key and an empty string as it's value. I also deleted the line that sets the empty string if the Host header contains no port reference.

boazsegev commented:
Thanks for the updates!

I hope to look into this in the next few days, however, I am slightly worried that a ` rb_hash_aset(env, SERVER_PORT,...)` could cause a re-hashing of the keys and memory allocations in the hash map. 

Since this will likely be performed for every request and within the GIL, it could create a performance penalty for setups that cause clients to use `host:port` (is this common?).

This was the reason for creating the `env` template and pre-initializing the common keys - avoiding the re-hashing algorithm as keys are added.

On the other hand, `rb_hash_delete(env, SERVER_PORT)` shouldn't cause a rehashing, but it would invoke a smaller performance penalty for the regular `host` headers that have no `:port` information (which might be better for most use cases).

Anyway, I'll have a look at the code later and think it over.

Cheers!

boazsegev commented:
I was sure I already accepted this... instead it was just waiting in the backlog, quietly snoozing between the cracks...

Thanks again for the PR!

adam12 commented:
:joy: I didn't want to pester. Life of the OSS developer and all that.

Hope all is well with you. Cheers!

yildizsc commented:
Thank you very much @boazsegev @adam12
